File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/ledger/validations/HouseholdSharedAccountValidator.java
package com.decrux.pocketr.api.services.ledger.validations;

import com.decrux.pocketr.api.entities.db.ledger.Account;
import com.decrux.pocketr.api.exceptions.ForbiddenException;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Set;
import java.util.UUID;

@Component
public class HouseholdSharedAccountValidator {

    public void validate(List<Account> nonOwnedAccounts, Set<UUID> sharedAccountIds) {
        for (Account account : nonOwnedAccounts) {
            UUID accountId = requireNotNull(account.getId());
            if (!sharedAccountIds.contains(accountId)) {
                throw new ForbiddenException("Account '" + account.getName() + "' is not shared into household");
            }
        }
    }

    private static <T> T requireNotNull(T value) {
        if (value == null) {
            throw new IllegalArgumentException("Required value was null.");
        }
        return value;
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr/api/services/ledger/validations/HouseholdSharedAccountValidator.kt
package com.decrux.pocketr.api.services.ledger.validations

import com.decrux.pocketr.api.entities.db.ledger.Account
import com.decrux.pocketr.api.exceptions.ForbiddenException
import com.decrux.pocketr.api.services.household.ManageHousehold
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class HouseholdSharedAccountValidator {
    fun validate(
        nonOwnedAccounts: List<Account>,
        manageHousehold: ManageHousehold,
        householdId: UUID,
    ) {
        nonOwnedAccounts.forEach { account ->
            val accountId = requireNotNull(account.id)
            if (!manageHousehold.isAccountShared(householdId, accountId)) {
                throw ForbiddenException(
                    "Account '${account.name}' is not shared into household",
                )
            }
        }
    }
}
